### ISSUE TITLE
BOM issue with μ character

### ISSUE THREAD

**Issue by harikt** [duncan3dc/bom-string#1]
Hi,

Thank you for the library. 

I came across an issue trying to [import a csv file](http://www.data-explorer.com/content/data/standard-model-of-particle-physics-csv.zip) . It is utf-16 encoded file.

The first issue was with `ID`, and this library was able to fix the same. When iterating across around at 9th row 

![image](https://cloud.githubusercontent.com/assets/120454/18999763/db02b96a-875c-11e6-847f-d526ddc3da46.png)

the `Symbol` `μ` was not converted correctly. I am not sure what is the real reason. In case you know something let me know.

Thank you.

**Comment by duncan3dc** [comment]
Hi @harikt, when I open that file with a BOM aware editor, the symbol shows as `μ`, so I'm not sure there's any issue here.
Do you believe that symbol is incorrect? If so, what should it be?

**Comment by harikt** [comment]
Hi @duncan3dc ,

Thanks for your quick reply. I am not sure what is the correct symbol for the same.

```
array(21) {
  ["ID"]=>
  string(3) "9"
  ["Name"]=>
  string(9) "Muon"
  ["Symbol"]=>
  string(3) "��"
 ....
}
```

Something like the above was thrown when reading from the file via `fgetcsv` function. You can see the symbol is not converted / fixed with the BOM filter I believe.

I may be wrong.

**Comment by duncan3dc** [comment]
Can you send me a link to a gist of your script? That file is working fine for me:

``` php
$data = file_get_contents(__DIR__ . "/physics.csv");
$clean = \duncan3dc\Bom\Util::removeBom($data);
foreach (explode("\n", $clean) as $line) {
    print_r(str_getcsv($line));
}
```

```
Array
(
    [0] => 9
    [1] => Muon
    [2] => μ
    [3] => Fermion
...
)
```

**Comment by harikt** [comment]
Your way of getting all data and escaping works nicely.

But this is something how I did, 

``` php
<?php
require 'vendor/autoload.php';

use duncan3dc\Bom\Util;

$file = __DIR__ . "/physics.csv";
$fp = fopen($file, "r");
while ($line = fgets($fp)) {
    $clean = Util::removeBom($line);
    $columns = str_getcsv($clean);
    print_r($columns);
}
```

I have also tried with `fgetcsv` , and going through each column value and trying to remove the BOM character. But that was also a failure.

Thank you.

**Comment by duncan3dc** [comment]
Ah ok, that won't work as the BOM is only at the start of the entire file, so once you strip it from line 1, the other lines are unaffected, and there's no way of knowing they are UTF-16.

You need to handle the BOM on a complete file basis. Would a `removeBomFromFile()` method be useful?

``` php
$file = __DIR__ . "/physics.csv";
Util::removeBomFromFile($file);
$fp = fopen($file, "r");
while ($row = fgetcsv($fp)) {
    print_r($row);
}
```

**Comment by harikt** [comment]
Oh I wasn't aware BOM is for the complete file. So how does `removeBomFromFile` going to work? Does it save back the cleaned text to the same file ?

**Comment by duncan3dc** [comment]
Yes that's right, it would overwrite the file with a UTF-8 version and no BOM, would that be useful?

**Comment by duncan3dc** [comment]
Rather than overwriting files which might be a bit presumptuous, I've created a stream filter instead:

``` php
use duncan3dc\Bom\StreamFilter;
stream_filter_register("bom-filter", StreamFilter::class);

$file = fopen($filename, "r");
stream_filter_append($file, "bom-filter");
while ($row = fgetcsv($file)) {
    print_r($row);
}
```

This is available in [0.2.0](https://github.com/duncan3dc/bom-string/releases/tag/0.2.0)

**Comment by harikt** [comment]
Thank you @duncan3dc . It also opened a few new functions and how it can be used to me.
